### DISCUSSION TITLE
How to convert the below filter-branch command ?

### DISCUSSION THREAD

**Discussion by DilLip-Chowdary-Codes** [newren/git-filter-repo#471] [Q&A]
Checked [docs](https://github.com/newren/git-filter-repo/blob/main/Documentation/converting-from-filter-branch.md#cheat-sheet-conversion-of-examples-from-the-filter-branch-manpage) but was not able to find a way

```bash
git filter-branch --commit-filter 'if [ $(git show --no-patch --no-notes --pretty='%ct' $GIT_COMMIT) -ge 1682879400 ]; then git commit-tree "$@"; else skip_commit "$@"; fi' -- --all
```

**Comment by andry81** [comment]
Looks like the command tries to cut off all commits younger than a commit date-time retaining all files and changes in the first commit in each branch. You can just find all these commits manually and if there is not many of them, then just cut off all the parents in each commit through:

```
git replace --graft commit
git filter-repo --force
git for-each-ref --format='delete %(refname)' refs/replace | git update-ref --stdin
```

**Reply by DilLip-Chowdary-Codes** [reply]
Yeah, basically we trying to [solve this problem](https://stackoverflow.com/questions/76338041/git-squash-all-commits-till-specified-date-into-one-commit) is it possible with our `filter-repo` without any manual work?

At least by modifying the python code in git-filter-repo, is there any possibility?

**Reply by andry81** [reply]
This is what you actually need to do on your own. Go over all branches and find the first commit:

```shell
IFS=$'\r\n'
for show_ref_line in `git show-ref --heads`; do
  IFS=$' \t'; read prefix head_ref <<< $show_ref_line
  echo $head_ref
  git log --until=2022-10-28T -1 --format='%H' $head_ref
done
```

And run replace command on each commit.

But, there is a catch. If you have a merge commits between branches, then `git log` can return for different branches the same commit hash.
To workaround that you can change the time before the rewritten commit to select different branch commit to cut off before it.

**Comment by newren** [comment]
Untested, but if you add `--full-tree` to the list of options that filter-repo passes to fast-export (search for `--show-original-ids` and add `--full-tree` to that list), then I _think_ you could translate your filter-branch command to the following:
```
git filter-repo --commit-callback '
    epoch = int(commit.committer_date.split()[0])
    if epoch < 1682879400:
        commit.skip()
'
```
